projecttacoma/deqm-test-server-frontend
# Create resource request

## Summary
Implementation of the create new resource submit button. POST request with the code editor's contents is sent to the server when the submit button is clicked.

## New behavior
In any /[resourceType]/create, when the submit button is clicked, a POST request is sent to the test server. 
- If a resource is successfully created:

  1.    a **success notification** appears with the new resource's ID as a Link that navigates to the [resourceType]/[id] page displaying the new resource's body 

  1.   and the user is **redirected** to the resourceType's home page.

- If the request is unsuccessful (server returns 401, fetch throws an error, etc.) an error message notification appears on the create new resource page.
 
## Code changes
- pages/create.tsx, particularly the editorSubmitHandler function
-  tests/pages/create.test.tsx, where unit testing was attempted

## Testing Guidance
- In a .env.local file, include the following: `NEXT_PUBLIC_DEQM_SERVER=http://localhost:3000/4_0_1`. Then start the DEQM test server locally.
- To start the frontend: `npm run dev` then navigate to http://localhost:3001 in a browser
- Click on any resource type from the navigation menu on the left. Then click on the "Create New [resourceType]" button in the upper right hand corner of the main page body. Test the submit button implementation in the following 3 ways:

  1. Type into the code editor a valid fhir resource body to post.
  Ex: 
  ```
  { 
     "resourceType": "Patient",
     "id": "JaneDoe"
  }
  ```
  then click submit. User should be redirected to the resourceType's home page. A green check-marked success notification should appear with the new resource's ID as a Link that navigates to a page displaying the resource's body.
   

  2. Type `{}` into the code editor, then click submit. A red X-marked error notification will appear with an error message
  3. Disconnect from the server, then enter {} or other valid JSON into the code editor, then click submit. A red X-marked error notification will appear with a "Problem connecting to server" message.
- **NOTE** that the implementation of the POST request has NOT been unit tested because there were issues encountered when trying to mock a user inputting valid JSON content to the code editor component in order to enable the Submit button to send the request. 
  - A test in __tests__/pages/create.test.tsx is **purposefully skipped** so that the test framework is preserved for investigation down the line.
